### PR TITLE
NERSC Changes

### PR DESCRIPTION
**Pull request by becker33**
Do not merge.

This will be factored into several commits and merged separately, this is to keep track of changes.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+build:
+  stage: build
+  tags:
+    - nersc
+    - edison
+  script:
+    - echo "Project $CI_PROJECT_NAME, branch $CI_COMMIT_REF_NAME"
+    - echo "About to run spack"
+    - spack --help
+    - echo "...did that work...?"
```
Comment (becker33) [inline .gitlab-ci.yml]: Ignore, will not be merged with mainline.

File: etc/spack/defaults/config.yaml [hunk 2]
```diff
@@ -74,4 +78,4 @@ config:
   # The default number of jobs to use when running `make` in parallel.
   # If set to 4, for example, `spack install` will run `make -j4`.
   # If not set, all available cores are used by default.
-  # build_jobs: 4
+  build_jobs: 8
```
Comment (becker33) [inline etc/spack/defaults/config.yaml]: Ignore, will not be merged with mainline.

File: etc/spack/defaults/cray/packages.yaml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        modules:
+            papi@5.5.1.3%gcc: papi
+            papi@5.5.1.3%intel: papi
+            papi@5.5.1.3%cce: papi
+    trilinos:
+        buildable: false
+        modules:
+            trilinos@12.10.1.1%gcc: cray-trilinos
+            trilinos@12.10.1.1%intel: cray-trilinos
+            trilinos@12.10.1.1%cce: cray-trilinos
```
Comment (becker33) [inline etc/spack/defaults/cray/packages.yaml]: Ignore, will not be merged with mainline

File: lib/spack/spack/build_environment.py [hunk 4]
```diff
@@ -70,7 +70,8 @@
 from spack.environment import EnvironmentModifications, validate
 from spack.util.environment import env_flag, filter_system_paths, get_path
 from spack.util.executable import Executable
-from spack.util.module_cmd import load_module, get_path_from_module
+from spack.util.module_cmd import (load_module, get_path_from_module,
+                                   unload_module)
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]: remove `unload_module`

File: lib/spack/spack/build_environment.py [hunk 5]
```diff
@@ -337,6 +338,13 @@ def set_build_environment_variables(pkg, env, dirty):
             if os.path.isdir(pcdir):
                 env.prepend_path('PKG_CONFIG_PATH', pcdir)
 
+    if pkg.architecture.target.module_name:
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]: remove, duplicates later load.

File: lib/spack/spack/build_environment.py [hunk 6]
```diff
@@ -337,6 +338,13 @@ def set_build_environment_variables(pkg, env, dirty):
...
+        load_module(pkg.architecture.target.module_name)
+    if "cray" in pkg.architecture.platform.name:
```
Comment (becker33) [inline lib/spack/spack/build_environment.py]: Move to cray.setup_platform_environment()

Also unload `cray-mpich` so we don't accidentally find MPI when we didn't list it as a dependency.

File: lib/spack/spack/cmd/flake8.py [hunk 7]
```diff
@@ -114,6 +114,8 @@ def changed_files(args):
     """Get list of changed files in the Spack repository."""
 
     git = which('git', required=True)
+    current_branch = git("rev-parse", "--abbrev-ref", "HEAD",
```
Comment (becker33) [inline lib/spack/spack/cmd/flake8.py]: remove, there is now a -b option.

File: lib/spack/spack/compiler.py [hunk 8]
```diff
@@ -214,6 +214,9 @@ def cxx17_flag(self):
             "If you think it should, please edit the compiler subclass and",
             "submit a pull request or issue.")
 
+    @property
```
Comment (becker33) [inline lib/spack/spack/compiler.py]: Not mainlined

File: lib/spack/spack/concretize.py [hunk 9]
```diff
@@ -218,9 +218,12 @@ def concretize_architecture(self, spec):
         sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
         spec_changed = False
 
-        if spec.architecture is None:
+        if spec.architecture is None and root_arch is None:
```
Comment (becker33) [inline lib/spack/spack/concretize.py]: Will not mainline, major fix in develop

File: lib/spack/spack/parse.py [hunk 10]
```diff
@@ -162,7 +161,10 @@ def expect(self, id):
 
     def setup(self, text):
         if isinstance(text, string_types):
-            text = shlex.split(text)
+            if sys.version_info >= (2, 7, 9):
```
Comment (becker33) [inline lib/spack/spack/parse.py]: Fixed another way in develop, ignore

File: lib/spack/spack/platforms/cray.py [hunk 11]
```diff
@@ -103,8 +109,15 @@ def setup_platform_environment(cls, pkg, env):
         """ Change the linker to default dynamic to be more
             similar to linux/standard linker behavior
         """
-        env.set('CRAYPE_LINK_TYPE', 'dynamic')
+        packages = spack.config.get_config("packages")
```
Comment (becker33) [inline lib/spack/spack/platforms/cray.py]: This needs to be superseded by some sort of universal variant for `shared` or `link_type`.

File: lib/spack/spack/schema/packages.py [hunk 12]
```diff
@@ -82,6 +82,10 @@
                                 {'type': 'array',
                                  'items': {'type': 'string'}}],
                         },
+                        'craype_link_type': {
```
Comment (becker33) [inline lib/spack/spack/schema/packages.py]: Ignore

File: lib/spack/spack/spec.py [hunk 13]
```diff
@@ -3033,6 +3033,19 @@ def write(s, c):
                             fmt % token_transform(str(self.architecture)),
                             '='
                         )
+                elif named_str == "PLATFORM":
```
Comment (becker33) [inline lib/spack/spack/spec.py]: Merge to pull these to top level.

File: lib/spack/spack/util/module_cmd.py [hunk 14]
```diff
@@ -115,6 +115,14 @@ def get_module_cmd_from_bash(bashopts=''):
     return module_cmd
 
 
+def unload_module(mod):
+    """Takes a module name and unloads the module from the environment. It does
```
Comment (becker33) [inline lib/spack/spack/util/module_cmd.py]: Part of above PR with `unload_module` in `build_environment.py`

File: lib/spack/spack/util/path.py [hunk 15]
```diff
@@ -39,6 +39,9 @@
     'spack': spack.prefix,
     'user': getpass.getuser(),
     'tempdir': tempfile.gettempdir(),
+    'home': os.environ.get("HOME", ""), # $home and $scratch are NERSC specific
+    'scratch': os.environ.get("SCRATCH", ""),
```
Comment (becker33) [inline lib/spack/spack/util/path.py]: Ignore

File: etc/spack/defaults/config.yaml [hunk 16]
```diff
@@ -16,12 +16,15 @@
...
   # This is the path to the root of the Spack install tree.
   # You can use $spack here to refer to the root of the spack instance.
-  install_tree: $spack/opt/spack
+  install_tree: /global/common/sw   # Change to the desired installation
 
   # Locations where templates should be found
   template_dirs:
     - $spack/templates
 
+  # This is the scheme that will be the default for NERSC.
```
Comment (becker33) [inline etc/spack/defaults/config.yaml]: We'll add this section with change that uses it.

File: lib/spack/spack/util/path.py (hunk 15 above)
Comment (mamelara) [reply lib/spack/spack/util/path.py]: i think SUSER was more for us at NERSC

**Comment by alalazo** [comment]
@becker33 Is it fine to close this PR or is there anything that still needs to be factored out?

**Comment by alalazo** [comment]
@becker33 Do we still need this PR, or can we close it?

**Pull request closed without merging**